### PR TITLE
Fix time zone rounding edge case for DST overlaps

### LINKED ISSUE #20833: TimeZoneRoundingTests#testRoundingRandom fails reproducibly
```
 2> REPRODUCE WITH: gradle :core:test -Dtests.seed=17AC5B1BB751F423 -Dtests.class=org.elasticsearch.common.rounding.TimeZoneRoundingTests -Dtests.method="testRoundingRandom" -Dtests.security.manager=true -Dtests.locale=ko -Dtests.timezone=Indian/Comoro
FAILURE 0.69s | TimeZoneRoundingTests.testRoundingRandom <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: dateBetween should round down to roundedDate
   > Expected: Expected: 2024-10-27T00:00:00.000Z [1729987200000] 
   >      but:  was "2024-10-27T00:00:00.000-01:00 [1729990800000]"
   >    at __randomizedtesting.SeedInfo.seed([17AC5B1BB751F423:B8E2E3098394017E]:0)
   >    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   >    at org.elasticsearch.common.rounding.TimeZoneRoundingTests.assertInterval(TimeZoneRoundingTests.java:541)
   >    at org.elasticsearch.common.rounding.TimeZoneRoundingTests.testRoundingRandom(TimeZoneRoundingTests.java:195)
   >    at java.lang.Thread.run(Thread.java:745)
  2> NOTE: leaving temporary files on disk at: /Users/simon/projects/elasticsearch/core/build/testrun/test/J0/temp/org.elasticsearch.common.rounding.TimeZoneRoundingTests_17AC5B1BB751F423-001
  2> NOTE: test params are: codec=Asserting(Lucene62): {}, docValues:{}, maxPointsInLeafNode=386, maxMBSortInHeap=5.970189533938357, sim=ClassicSimilarity, locale=ko, timezone=Indian/Comoro
  2> NOTE: Mac OS X 10.11.6 x86_64/Oracle Corporation 1.8.0_66 (64-bit)/cpus=4,threads=1,free=434107792,total=514850816
  2> NOTE: All tests run in this JVM: [TimeZoneRoundingTests]
Completed [1/1] in 2.26s, 1 test, 1 failure <<< FAILURES!
```

### PR DESCRIPTION
**Pull request by cbuescher**
When using TimeUnitRounding with a DAY_OF_MONTH unit, failing tests in #20833 uncovered an issue when the DST shift happenes just one hour after midnight local time and sets back the clock to midnight, leading to an overlap. Previously this would lead to two different rounding values, depending on whether a date before or after the transition was rounded. This change detects this special case and correct for it by using the previous rounding date for both cases.

In this particular case (tz: "Atlantic/Azores", dates around the DST transition on e.g. 2000-10-29T01:00:00.000Z, unit: DAY_OF_MONTH) we currently get the overlapping part during the DST transition as a separate rounding interval as illustrated in this table:

| date                                      | round(date)                          | nextRoundingValue(date)      |
| --------------------------------|---------------------------------|---------------------------------|
| 2000-10-28T22:00:00.000Z | 2000-10-28T00:00:00.000Z | 2000-10-29T00:00:00.000Z |
| 2000-10-28T23:00:00.000Z | 2000-10-28T00:00:00.000Z | 2000-10-29T00:00:00.000Z |
| 2000-10-29T00:00:00.000Z | **2000-10-29T00:00:00.000Z** | 2000-10-30T00:00:00.000-01:00 |
| 2000-10-29T00:00:00.000-01:00 | **2000-10-29T00:00:00.000-01:00** | 2000-10-30T00:00:00.000-01:00 |
| 2000-10-29T01:00:00.000-01:00 | 2000-10-29T00:00:00.000-01:00 | 2000-10-30T00:00:00.000-01:00 |
| 2000-10-29T02:00:00.000-01:00 | 2000-10-29T00:00:00.000-01:00 | 2000-10-30T00:00:00.000-01:00 |

According to https://www.timeanddate.com/time/change/portugal/horta?year=2000 the DST change happenes on Oct-29th at 1am local time, turning back the clock one hour (to offset -01:00). Currently the dates between "2000-10-29T00:00:00.000Z" and "2000-10-29T01:00:00.000Z" all round down to "2000-10-29T00:00:00.000Z" (before the transition) and the dates after but before next midnight round down to "2000-10-29T00:00:00.000-01:00". For a day rounding we would prefer a 25h hour bucket for "2000-10-29". 

With this fix, the situation above changes to 

| date                                      | round(date)                          | nextRoundingValue(date)      |
| --------------------------------|---------------------------------|---------------------------------|
| 2000-10-28T22:00:00.000Z | 2000-10-28T00:00:00.000Z | 2000-10-29T00:00:00.000Z |
| 2000-10-28T23:00:00.000Z | 2000-10-28T00:00:00.000Z | 2000-10-29T00:00:00.000Z |
| 2000-10-29T00:00:00.000Z | 2000-10-29T00:00:00.000Z | 2000-10-30T00:00:00.000-01:00 |
| 2000-10-29T00:00:00.000-01:00 | 2000-10-29T00:00:00.000Z | 2000-10-30T00:00:00.000-01:00 |
| 2000-10-29T01:00:00.000-01:00 | 2000-10-29T00:00:00.000Z | 2000-10-30T00:00:00.000-01:00 |
| 2000-10-29T02:00:00.000-01:00 | 2000-10-29T00:00:00.000Z | 2000-10-30T00:00:00.000-01:00 |

So now everything on "2000-10-29" gets rounded down to "2000-10-29T00:00:00.000Z".

Closes #20833

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/rounding/Rounding.java [hunk 1]
```diff
@@ -128,20 +128,40 @@ public byte id() {
...
 
+        /**
+         * Check if the current time instant is at a start of a DST overlap by
+         * comparing the offset of the instant and the previous millisecond. We
+         * want to detect negative offset changes that result in an overlap
+         */
+        private boolean onDSTOverlap(long instant) {
+            int offset = timeZone.getOffset(instant);
+            int offsetTicBefore = timeZone.getOffset(instant - 1);
+            return ((offset != offsetTicBefore) && (offset - offsetTicBefore < 0));
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/rounding/Rounding.java]: couldn't this be just `return offset < offsetTicBefore;`?

File: core/src/main/java/org/elasticsearch/common/rounding/Rounding.java [hunk 2]
```diff
@@ -128,20 +128,40 @@ public byte id() {
...
-                rounded = field.roundFloor(rounded);
+            if (timeZone.isFixed() == false) {
+                if (timeZone.getOffset(utcMillis) != timeZone.getOffset(rounded)) {
+                    // in this case, we crossed a time zone transition. In some
+                    // edge cases this will result in a value that is not a rounded
+                    // value itself. We need to round again to make sure. This will have no effect in
+                    // cases where 'rounded' was already a proper rounded value
+                    rounded = field.roundFloor(rounded);
+                } else {
+                    if (onDSTOverlap(rounded)) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/rounding/Rounding.java]: Skipping intervals that are less that the field duration sounds good to me regardless of whether we are on a DST overlap, so I'm wondering whether we want to get rid of this `if` statement (assuming there are no performance implications)?

**Review by jpountz — COMMENTED** [review]
I left a couple questions.

File: core/src/main/java/org/elasticsearch/common/rounding/Rounding.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/rounding/Rounding.java]: sure

File: core/src/main/java/org/elasticsearch/common/rounding/Rounding.java (hunk 2 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/rounding/Rounding.java]: I think generally "skipping" (or merging, like here) intervals that are less than the field duration is this is not always the right thing. Imagine - for example - a day that due to DST shift has only 23h (typically on DST start). We don't want to join that with the previous or the following day. I can look into this once again but its very likely some of our existing tests blow up when doing this.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/rounding/Rounding.java]: I just tried removing this check and it fails various test cases, so I think this supports the abover reasoning

**Comment by cbuescher** [comment]
@jpountz thanks for the review, I hope I answered your question and just pushed another commit addressing your other comment. Since this is not a very prominent problem I'm not sure this should be part of the next patch release, wdyt?

**Review by jpountz — APPROVED** [review]
Perfect, LGTM.

**Comment by cbuescher** [comment]
Merged on 5.x with 5846c7bb

**Pull request merged**